File: VB/Drag Drop rows/CustomGridDragDropManager.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows
Imports System.Windows.Input
Imports DevExpress.Xpf.Grid

Namespace Drag_Drop_rows
	Public Class CustomGridDragDropManager
		Inherits GridDragDropManager
		Implements IDragRowVisibilitySupport
		Public Class CustomGridDragSource
			Inherits GridDragSource
			Public Sub New(ByVal dragDropManager As DragDropManagerBase)
				MyBase.New(dragDropManager)
			End Sub

			Protected Overrides Function GetDraggingRows(ByVal e As MouseButtonEventArgs) As System.Collections.IList
				CustomGridDragDropManager.SetDraggingRowsData()
				Return MyBase.GetDraggingRows(e)
			End Function

			Private ReadOnly Property CustomGridDragDropManager() As CustomGridDragDropManager
				Get
					Return CType(MyBase.dragDropManager, CustomGridDragDropManager)
				End Get
			End Property
		End Class

		Private privateDraggingRowsData As List(Of Object)
		Public Property DraggingRowsData() As List(Of Object) Implements IDragRowVisibilitySupport.DraggingRowsData
			Get
				Return privateDraggingRowsData
			End Get
			Set(ByVal value As List(Of Object))
				privateDraggingRowsData = value
			End Set
		End Property

		Protected Overrides Function CreateDragSource(ByVal dataViewDragDropManager As DataViewDragDropManager) As DevExpress.Xpf.Core.ISupportDragDrop
			Return New CustomGridDragSource(dataViewDragDropManager)
		End Function

		Protected Overrides Function CreateDragElement(ByVal offset As Point, ByVal owner As FrameworkElement) As DevExpress.Xpf.Core.IDragElement
			Return New CustomDataControlDragElement(Me, offset, owner)
		End Function

		Public Sub SetDraggingRowsData()
			Dim list As New List(Of Object)()
            Dim view As TableView = TryCast(MyBase.View, TableView)
			For Each rowHandle As Integer In view.GetSelectedRowHandles()
				list.Add((TryCast(view.GetRowElementByRowHandle(rowHandle), GridRow)).RowDataContent.DataContext)
			Next rowHandle
			DraggingRowsData = list
		End Sub
	End Class
End Namespace

File: CS/Drag Drop rows/CustomGridDragDropManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Input;
using DevExpress.Xpf.Grid;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace Drag_Drop_rows {
    public class CustomGridDragDropManager : GridDragDropManager, IDragRowVisibilitySupport {
        public List<object> DraggingRowsData { get; set; }
        public TableView TableView { get { return View as TableView; } }
        protected override DevExpress.Xpf.Core.IDragElement CreateDragElement(Point offset, FrameworkElement owner) {
            return new CustomDataControlDragElement(this, offset, owner);
        }
        public void SetDraggingRowsData() {
            DraggingRowsData = GetRowImagesHelper.GetRowImages(TableView);
        }
        protected override System.Collections.IList CalcDraggingRows(DevExpress.Xpf.Core.IndependentMouseEventArgs e) {
            SetDraggingRowsData();
            return base.CalcDraggingRows(e);
        }
    }
    public static class GetRowImagesHelper {
        public static List<object> GetRowImages(DataViewBase view) {
            List<object> images = new List<object>();
            foreach (int rowHandle in view.DataControl.GetSelectedRowHandles()) {
                FrameworkElement element = view.GetRowElementByRowHandle(rowHandle);
                if (element != null) {
                    element = VisualTreeHelper.GetChild(element, 0) as FrameworkElement;
                    RenderTargetBitmap bmp = new RenderTargetBitmap((int)element.ActualWidth, (int)element.ActualHeight, 96, 96, PixelFormats.Pbgra32);
                    bmp.Render(element);
                    images.Add(new Image() { ImageSource = bmp, Width = element.ActualWidth, Height = element.ActualHeight });
                }
            }
            return images;
        }
    }
    public class Image {
        public ImageSource ImageSource { get; set; }
        public double Width { get; set; }
        public double Height { get; set; }
    }
}